src/peripherals/atecc608a-tnglora-se/atecc608a-tnglora-se-hal.c

```diff
@@ -272,7 +271,7 @@ ATCA_STATUS hal_i2c_wake(ATCAIface iface)
 ATCA_STATUS hal_i2c_idle(ATCAIface iface)
 {
     uint8_t buffer[1] = { 0x2 }; // idle word address value 
-    I2cMcuWriteBuffer((I2c_t*)NULL, iface->mIfaceCFG->atcai2c.slave_address,0, buffer, (size_t)1); 
+    I2cMcuWriteBuffer((I2c_t*)NULL, iface->mIfaceCFG->atcai2c.slave_address,0x2, buffer, (size_t)1);
```

Comment:
Can you please explain this line as I think it is incorrect.
Here is why:
- You are sending `0x2` `0x2` instead of the original `0x0` `0x2` .
- If you send an empty buffer,`0x2 0x0`, you are breaking the SE data sheet operation of the idle command.

All in all I think it is ok to send a reset before `0x2`

Reply:
> You are sending 0x2 0x2 instead of the original 0x0 0x2 .

If you send anything (independent of the data buffer) to address 0x0, you reset the device. It doesn't make sense to reset the device first and then send a 0x2, because the device won't know what to do with 0x2 after it's reset. But I think this discussion is highly related to the discussion above, where we talk about compatibility between different boards and why we put the data address in the first position of the data buffer or if we should solve this issue somewhere else in the driver of the SAMR34. 

> If you send an empty buffer,0x2 0x0, you are breaking the SE data sheet operation of the idle command.


You're right, it's not ideal to send a data buffer of 0x2 to the data address 0x2. But what I observed in the low level drivers of my mcu (LPC55S16) and in [stm32l4xx_hal_i2c.c](https://github.com/Lora-net/LoRaMac-node/blob/118771fee14d5cfb3edbe4b0846a404df9002443/src/boards/mcu/stm32/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_i2c.c#L2351), that the drivers don't send a stop condition if the data buffer's size is not bigger than 0. And the expected sequence is as followed stated in the data sheet: "The idle sequence consists of the proper device address followed by the value of 0x02 as the word address followed by a Stop condition."

> All in all I think it is ok to send a reset before 0x2

Well, I tried that as well (also with a Logic Analyzer to see what happens) and just addressed 0x0 with the data buffer 0x2, but that doesn't work either. This resets the device, that's not idle...

Reply:
Multiple issues here:
> Well, I tried that as well (also with a Logic Analyzer to see what happens) and just addressed 0x0 with the data buffer 0x2, but that doesn't work either. This resets the device, that's not idle...
 
Can you please explain what you observe?

Here is a snippet from the data sheet, your interpretation of the word reset  might be different from what the data sheet shows.
![image](https://user-images.githubusercontent.com/24204263/98521313-ebf5ac00-227b-11eb-9603-9b3dc1149503.png)


It is clear from my points and yours that the current LMN implementation where you enforce an added data address, is the cause of the issue, and your fix is also problematic, because you are trying to add an improvement and compromise on functionality.

ST has a solution for this where they provide another implementation for the underlying I2C driver instead of `HAL_I2C_Mem_Write`
```c
/**
  * @brief  Transmits in master mode an amount of data in blocking mode.
  * @param  hi2c Pointer to a I2C_HandleTypeDef structure that contains
  *                the configuration information for the specified I2C.
  * @param  DevAddress Target device address: The device 7 bits address value
  *         in datasheet must be shifted to the left before calling the interface
  * @param  pData Pointer to data buffer
  * @param  Size Amount of data to be sent
  * @param  Timeout Timeout duration
  * @retval HAL status
  */
HAL_StatusTypeDef HAL_I2C_Master_Transmit(I2C_HandleTypeDef *hi2c, uint16_t DevAddress, uint8_t *pData, uint16_t Size,
                                          uint32_t Timeout)
```


So why are you ok with sending `0x2` `0x2` to be compatible with the incorrectly used ST API?
Instead you assume if the transmission size is only 1 byte, that you should pass the buffer as its is without concatenating and address to it.

This also means that the current implementation where a buffer with one byte is sent is the correct one and the SAMR34 implementation is the correct one when it comes to sending one byte.

Reply:
I see your point, my proposal is not ideal either, but the current code doesn't work. 

> Can you please explain what you observe?

If I send `0x0` `0x2`, the secure element doesn't idle. Because of this, the secure element won't reply as expected when sending a wake up signal later. Have a look at the following Logic Analyzer Output. First, how it looks with sending `0x2` `0x2` for `idle`:
```
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x04 0x11 0x33 0x43
write to 0x59 ack data: 0x03 0x07 0x02 0x80 0x00 0x00 0x09 0xAD 
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23 0x01 0x23 0x4B 0x16 0x00 0x00 0x60 0x02 0x0F 0x3A 0x7D 0xF2 0x27 0x01 0x55 0x00 0xB2 0x00 0x00 0x01 0x8F 0xBD 0x85 0x8F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0x6B 0x82
write to 0x59 ack data: 0x02 0x02
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x04 0x11 0x33 0x43
```
We see the proper wake up on the last two lines, with the expected answer from the secure element (definition in table 4-3 page 21 in the [datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/ATECC608A-for-LoRaWAN-Data-Sheet-DS40002154B.pdf#page=21&zoom=auto,-663,710)). Now, if I send `0x0` `0x2` for `idle`:
```
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x04 0x11 0x33 0x43
write to 0x59 ack data: 0x03 0x07 0x02 0x80 0x00 0x00 0x09 0xAD 
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23 0x01 0x23 0x4B 0x16 0x00 0x00 0x60 0x02 0x0F 0x3A 0x7D 0xF2 0x27 0x01 0x55 0x00 0xB2 0x00 0x00 0x01 0x8F 0xBD 0x85 0x8F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0x6B 0x82
write to 0x59 ack data: 0x00 0x02 
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23 0x01 0x23 0x4B
```
We see on the last two lines, that the secure element was not in idle and doesn't reply as expected to the wake up signal. 

> ST has a solution for this where they provide another implementation for the underlying I2C driver instead of HAL_I2C_Mem_Write

I guess this could work. But in order to have a proper abstraction, one would have to change all i2c-board.c implementations of all the boards or add the possibility in i2c-board.h to have this method. Or how did you think to access this method from atecc608a-tnglora-se-hal.c? 

> So why are you ok with sending 0x2 0x2 to be compatible with the incorrectly used ST API?
Instead you assume if the transmission size is only 1 byte, that you should pass the buffer as its is without concatenating and address to it.

As described above with the output of the Logic Analyzer: Without explicitly addressing `0x2` and just send it as a buffer to address `0x0` doesn't work. 

> This also means that the current implementation where a buffer with one byte is sent is the correct one and the SAMR34 implementation is the correct one when it comes to sending one byte.

You're right, in this case this would be an ideal implementation. But as discussed in the conversation above, the `I2cMcuWriteBuffer` of the SAMR34 doesn't comply to the header files of the rest of the stack. Since we are going to change this function anyway, this won't be the solution for this problem here.

## Now how to proceed

I see two possible ways for a simple and working implementation. 
1. Just addressing `0x2` without any data buffer, but in consequence leaving away the stop condition:
In the following Logic Analyzer output, we see that this method works as well. But as mentioned in the datasheet, leaving away the stop condition may eventually result in a loss of synchronization. With the current driver, the synchronization should be gained back automatically with the implemented retries and the wake up sequence as far as I can see. 
```
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x04 0x11 0x33 0x43
write to 0x59 ack data: 0x03 0x07 0x02 0x80 0x00 0x00 0x09 0xAD 
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x23 0x01 0x23 0x4B 0x16 0x00 0x00 0x60 0x02 0x0F 0x3A 0x7D 0xF2 0x27 0x01 0x55 0x00 0xB2 0x00 0x00 0x01 0x8F 0xBD 0x85 0x8F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0xCE 0x0F 0x6B 0x82
write to 0x59 ack data: 0x02 
write to 0x00 nak
write to 0x59 ack data: 0x00 
read to 0x59 ack data: 0x04 0x11 0x33 0x43
```
2. Addressing `0x2` with a data buffer containing something random (in this case `0x2`):
This doesn't comply to the [idle sequence described in the datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/ATECC608A-for-LoRaWAN-Data-Sheet-DS40002154B.pdf#page=62&zoom=auto,-663,380), but the datasheet doesn't mention what happens if we send actual data with this command. On the other hand, the secure element works as expected and we can comply with the usual definition of `I2cMcuWriteBuffer` in i2c-board.h. 

What do you think? IMHO I would go for option two.